Initial port of libsodium for Unikraft

mpp-opsy opened the pull request:
Add makefile and config for building libsodium on Unikraft. This
port builds the minimal configuration of soduium and only supports
aarch64.

Signed-off-by: Michalis Pappas <Michalis.Pappas@opensynergy.com>

Commit:
Initial port of libsodium for Unikraft

Add makefile and config for building libsodium on Unikraft. This
port builds the minimal configuration of soduium.

Signed-off-by: Michalis Pappas <michalis.pappas@opensynergy.com>

nderjung commented:
Hi,

Thanks for the PR! I have a few notes, see inline:

nderjung commented on Makefile.uk:
```diff
@@ -0,0 +1,201 @@
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+#  THIS HEADER MAY NOT BE EXTRACTED OR MODIFIED IN ANY WAY.
```
Please can you remove this line, it prevents future developers from adding their copyright if they make significant contributions to the file.  We have been combing through existing libraries and the core to remove this line from the license header.  [More detatils](https://github.com/unikraft/unikraft/commit/7fd6797bd5917acc515ef6ddbfa85621f4aacf5f).

mpp-opsy replied:
Oooops, fixed

nderjung commented on Makefile.uk:
```diff
@@ -0,0 +1,201 @@
+LIBSODIUM_VERSION=1.0.18
+LIBSODIUM_LIBRARY_VERSION_MAJOR=10
+LIBSODIUM_LIBRARY_VERSION_MINOR=3
+LIBSODIUM_URL=https://download.libsodium.org/libsodium/releases/libsodium-$(LIBSODIUM_VERSION)-stable.tar.gz
```
As discussed offline, can you replace this with the [Github repo](https://github.com/jedisct1/libsodium) equivilant link to the archive, for some reason this URL doesn't play nicely with our CI and is not starting the build-to-test sequence.

mpp-opsy replied:
Done

nderjung reviewed (commented):
Hi,

Thanks for the PR! I have a few notes, see inline:

nderjung commented on Config.uk:
```diff
@@ -0,0 +1,4 @@
+menuconfig LIBSODIUM
+	bool "libsodium"
```
```diff
-	bool "libsodium"
+	bool "libsodium: A modern, portable, easy to use crypto library."
```

nderjung commented on MAINTAINERS.md:
```diff
@@ -0,0 +1,11 @@
+Maintainers List
+================
+
+For notes on how to read this information, please refer to `MAINTAINERS.md` in
+the main Unikraft repository.
+
+	LIBSODIUM-UNIKRAFT
+	M:	Felipe Huici <felipe.huici@neclab.eu>
+	M:	Razvan Deaconescu <razvan.deaconescu@cs.pub.ro>
+	L:	minios-devel@lists.xen.org
+	F: *
```
This file can be deleted, we are about to deprecate this file from the Unikraft ecosystem :) We are working on a new governance model (and repo) which will superced this.

mpp-opsy replied:
Done

nderjung commented on Makefile.uk:
```diff
@@ -0,0 +1,199 @@
+# -DHAVE_SYS_RANDOM_H=1 -DHAVE_ALLOCA_H=1 -DHAVE_ALLOCA=1 -DHAVE_MMAP=1 -DHAVE_MLOCK=1 -DHAVE_MADVISE=1 -DHAVE_MPROTECT=1
+# -DHAVE_GETRANDOM=1 -DHAVE_GETENTROPY=1 -DHAVE_POSIX_MEMALIGN=1 -DHAVE_GETPID=1 -DHAVE_NANOSLEEP=1 -DHAVE_EXPLICIT_BZERO=1 
+#
+LIBSODIUM_DEFINES += -DPACKAGE_NAME=\"libsodium\" -DPACKAGE_TARNAME=\"libsodium\" -DPACKAGE_VERSION=\"1.0.18\" -DPACKAGE_STRING=\"libsodium\ 1.0.18\" \
+ -DPACKAGE_BUGREPORT=\"https://github.com/jedisct1/libsodium/issues\" -DPACKAGE_URL=\"https://github.com/jedisct1/libsodium\" -DPACKAGE=\"libsodium\" \
+ -DVERSION=\"1.0.18\" -DMINIMAL=1 -DHAVE_PTHREAD_PRIO_INHERIT=1 -DHAVE_PTHREAD=1 -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 \
+ -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -D__EXTENSIONS__=1 -D_ALL_SOURCE=1 \
+ -D_GNU_SOURCE=1 -D_POSIX_PTHREAD_SEMANTICS=1 -D_TANDEM_SOURCE=1 -DHAVE_C_VARARRAYS=1 -DTLS=_Thread_local -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" \
+ -DHAVE_SYS_MMAN_H=1 -DNATIVE_LITTLE_ENDIAN=1 -DHAVE_INLINE_ASM=1 -DHAVE_TI_MODE=1 -DASM_HIDE_SYMBOL=.hidden -DHAVE_WEAK_SYMBOLS=1 -DHAVE_ATOMIC_OPS=1  -DCONFIGURED=1
```
Can you break this up into multiple lines per definition? I think it will make it easier to read :-) E.g.:
```Makefile
LIBSODIUM_DEFINES += -DPACKAGE_NAME=\"libsodium\" \
                     -DPACKAGE_TARNAME=\"libsodium\" \
                     ...
```

Additionally, I notice that the version number is inline here, please can you add it dynamically, i.e.:

```diff
                     ...
-                    -DPACKAGE_VERSION=\"1.0.18\" \
+                    -DPACKAGE_VERSION=\"$(LIBSODIUM_VERSION)\" \
```

mpp-opsy replied:
Done. Also made the dependency to pthread-embedded optional.

nderjung reviewed (commented):
Hi, thanks for the changes! :) I have a few more comments.  I'll give the library a whirl in the next few days to see it build, as it looks like the CI/CD system doesn't build it since it depends on newlib and pthread-embedded.

nderjung commented on Makefile.uk:
```diff
@@ -0,0 +1,233 @@
+LIBSODIUM_DEFINES += -DPACKAGE_NAME=\"libsodium\" \
+                     -DPACKAGE_TARNAME=\"libsodium\" \
+                     -DPACKAGE_VERSION=\"$(LIBSODIUM_VERSION)\" \
+                     -DPACKAGE_STRING=\"libsodium\ 1.0.18\" \
```
Can you please add the version number variable here too? :)

mpp-opsy replied:
Ooops, fixed now

michpappas committed:
Introduce libsodium tests

This commit introduces changes required for running libsodium's tests
on Unikraft. Tests can be enabled through the newly introduced kconfig
option CONFIG_LIBSODIUM_TESTS.

Minor changes on sodium's tests are introduced to allow building all
tests in a single Unikraft application. The core function that invokes
the test vectors in cmptest.h has been replaced in order to remove the
filesystem dependency for storing the test results.

The tests enabled are the ones executed when compiling libsodim natively
with --enable-minimal and running "make check". No test vectors have
been modified with the exception of sodium_core, sodium_util2, and
sodium_util3 to disable calls to exit() which cause the termination of
the Unikraft application.

Signed-off-by: Michalis Pappas <michalis.pappas@opensynergy.com>

michpappas commented:
Hi @nderjung, @razvand I just pushed an update with the following changes:

1. Fix an issue with sodium_init found by @samoilescusebastian (thanks for that!)
2. Update README.md
2. Introduce support for libsodium's tests

All tests that pass natively, also pass on the Unikraft port. Tested on linuxu (requires bumping memory to 16MiB), KVM on x86_64, and KVM on aarch64.

```
Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
                   Tethys 0.5.0~c1dc128
PASS: aead_aes256gcm
PASS: aead_aes256gcm2
PASS: aead_chacha20poly1305
PASS: aead_chacha20poly13052
PASS: aead_xchacha20poly1305
PASS: auth
PASS: auth2
PASS: auth3
PASS: auth5
PASS: auth6
PASS: auth7
PASS: box
PASS: box2
PASS: box7
PASS: box8
PASS: box_easy
PASS: box_easy2
PASS: box_seal
PASS: box_seed
PASS: chacha20
PASS: codecs
PASS: core1
PASS: core2
PASS: core3
PASS: core4
PASS: core5
PASS: core6
PASS: ed25519_convert
PASS: generichash
PASS: generichash2
PASS: generichash3
PASS: hash
PASS: hash3
PASS: kdf
PASS: keygen
PASS: kx
PASS: metamorphic
PASS: misuse
PASS: onetimeauth
PASS: onetimeauth2
PASS: onetimeauth7
PASS: pwhash_argon2i
PASS: pwhash_argon2id
PASS: randombytes
PASS: scalarmult
PASS: scalarmult2
PASS: scalarmult5
PASS: scalarmult6
PASS: scalarmult7
PASS: scalarmult8
PASS: secretbox
PASS: secretbox2
PASS: secretbox7
PASS: secretbox8
PASS: secretbox_easy
PASS: secretbox_easy2
PASS: secretstream
PASS: shorthash
PASS: sign
PASS: sodium_core
PASS: sodium_utils
PASS: sodium_utils2
PASS: sodium_utils2
PASS: sodium_version
PASS: stream
PASS: stream2
PASS: stream3
PASS: stream4
PASS: verify1
==============================================
Testsuite summary for libsodium 1.0.18
==============================================
# TOTAL: 69
# PASS: 69
# FAIL: 0
==============================================
```

razvand commented:
Thanks so much, @michpappas . Will look into it. @samoilescusebastian , please do a review.

samoilescusebastian reviewed (approved):
Looks fine to me.

Reviewed-by: Sebastian Samoilescu <samoilescusebastian@gmail.com>

razvand reviewed (approved):
All OK. Thanks, @michpappas 

Approved-by: Razvan Deaconescu <razvan.deaconescu@cs.pub.ro>

razvand commented:
Approved-by: Razvan Deaconescu <razvan.deaconescu@cs.pub.ro>

razvand reviewed (approved):
All OK.

Approved-by: Razvan Deaconescu <razvan.deaconescu@cs.pub.ro>
